include/deal.II/numerics/data_out_dof_data.templates.h

```diff
@@ -1709,32 +1676,107 @@ DataOut_DoFData<DoFHandlerType, patch_dim, patch_space_dim>::get_dataset_names()
+  // consecutive components.
+  //
+  // This sort of logic is also explained in some detail in
+  //   DataOut::build_one_patch().
+  for (const auto &input_data : dof_data)
+    if (input_data->is_complex_valued() == false ||
+        (input_data->postprocessor != nullptr))
+      {
+        for (unsigned int i = 0; i < input_data->names.size(); ++i)
+          names.push_back(input_data->names[i]);
+      }
+    else
+      {
+        // OK, so we have a complex-valued vector. We then need to go through
+        // all components and order them appropriately
+        for (unsigned int i = 0; i < input_data->names.size();
+             /* increment of i happens below */)
+          {
+            switch (input_data->data_component_interpretation[i])
+              {
+                case DataComponentInterpretation::component_is_scalar:
+                  {
+                    // It's a scalar. Just output real and imaginary parts one
+                    // after the other:
+                    names.push_back(input_data->names[i] + "_re");
+                    names.push_back(input_data->names[i] + "_im");
 
-  // Do the same as above for cell-type data
-  for (auto d = cell_data.begin(); d != cell_data.end(); ++d)
+                    // Move forward by one component
+                    ++i;
+
+                    break;
+                  }
+
+                case DataComponentInterpretation::component_is_part_of_vector:
+                  {
+                    // It's a vector. First output all real parts, then all
+                    // imaginary parts:
+                    const unsigned int size = patch_space_dim;
+                    for (unsigned int vec_comp = 0; vec_comp < size; ++vec_comp)
+                      names.push_back(input_data->names[i + vec_comp] + "_re");
+                    for (unsigned int vec_comp = 0; vec_comp < size; ++vec_comp)
+                      names.push_back(input_data->names[i + vec_comp] + "_im");
+
+                    // Move forward by dim components
+                    i += size;
+
+                    break;
+                  }
+
+                case DataComponentInterpretation::component_is_part_of_tensor:
+                  {
+                    // It's a tensor. First output all real parts, then all
+                    // imaginary parts:
+                    const unsigned int size = patch_space_dim * patch_space_dim;
+                    for (unsigned int tensor_comp = 0; tensor_comp < size;
+                         ++tensor_comp)
+                      names.push_back(input_data->names[i + tensor_comp] +
+                                      "_re");
+                    for (unsigned int tensor_comp = 0; tensor_comp < size;
+                         ++tensor_comp)
+                      names.push_back(input_data->names[i + tensor_comp] +
+                                      "_im");
+
+                    // Move forward by dim components
```

Comment:
```suggestion
                    // Move forward by dim*dim components
```
alternatively you could also just write `size`.